cli/cmd/identity.go

```diff
@@ -0,0 +1,231 @@
+package cmd
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/grantae/certinfo"
+	"github.com/linkerd/linkerd2/pkg/k8s"
+	"github.com/linkerd/linkerd2/pkg/k8s/resource"
+	tls1 "github.com/linkerd/linkerd2/pkg/tls"
+	"github.com/spf13/cobra"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+var emitLog bool
+
+type certificate struct {
+	pod         string
+	container   string
+	Certificate []*x509.Certificate
+	err         error
+}
+
+type identityOptions struct {
+	pod       string
+	namespace string
+	selector  string
+}
+
+func newIdentityOptions() *identityOptions {
+	return &identityOptions{
+		pod:       "",
+		namespace: "",
+		selector:  "",
+	}
+}
+
+func newCmdIdentity() *cobra.Command {
+	emitLog = false
+	options := newIdentityOptions()
+
+	cmd := &cobra.Command{
+		Use:   "identity [flags] (PODS)",
+		Short: "Display the certificate(s) of one or more selected pod(s)",
+		Long: `Display the certificate(s) of one or more selected pod(s).
+		
+		This command initiates a port-forward to a given pod or a set of pods and
+		fetches the tls certificate.
+		`,
+		Example: ` 
+		#Get certificate from pod foo-bar in the default namespace.
+		linkerd identity foo-bar
+		
+		#Get certificate from all pods with name=nginx
+		linkerd identity -l name=nginx
+		`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			k8sAPI, err := k8s.NewAPI(kubeconfigPath, kubeContext, impersonate, impersonateGroup, 0)
+			if err != nil {
+				return err
+			}
+
+			resources, err := resource.FetchKubernetesResources(cmd.Context(), k8sAPI, metav1.ListOptions{LabelSelector: k8s.ControllerNSLabel})
+			if err != nil {
+				return err
+			}
+			// check if linkerd control plane exists
+			if len(resources) == 0 {
+				return fmt.Errorf("Cannot find Linkerd\nValidate the install with: linkerd check")
+			}
+			if len(args) == 0 && options.selector == "" {
+				return fmt.Errorf("Provide the pod name argument or use the selector flag")
+			}
+
+			pods, err := getPods(cmd.Context(), k8sAPI, options.namespace, options.selector, args)
+			if err != nil {
+				return err
+			}
+
+			resultCerts := getCertificate(k8sAPI, pods, k8s.ProxyAdminPortName, emitLog)
+			for _, resultCert := range resultCerts {
+				if resultCert.err != nil {
+					fmt.Printf("\n%s", resultCert.err)
+					return nil
+				}
+				certChain := resultCert.Certificate
+				cert := certChain[len(certChain)-1]
+				result, err := certinfo.CertificateText(cert)
+				if err != nil {
+					fmt.Printf("\n%s", err)
+					return nil
+				}
+				fmt.Print(result)
+				fmt.Println("pem: |")
```

Comment:
Since we're already printing the entire certificate in an openssl-like format, I don't think we also need to print the PEM encoding.